### DISCUSSION TITLE
Error setting foreign keys directly

### DISCUSSION THREAD

**Discussion by pradyuman** [prisma/prisma#4410] [Q&A] [answered]
I get the following error when setting the `uncheckedScalarInputs` flag and trying to set the foreign key directly:
`Cannot read property 'minNumFields' of undefined`

For example, this query:

```ts
  await prisma.organizationUser.create({
    data: {
      roles: ["owner"],
      organization: {
        create: {
          name: "Organization",
          organizationPosts: {
            create: {
              name: "OrganizationPost",
              postId,
            },
          },
        },
      },
    },
    include: {
      organization: true,
      entity: true,
    },
  })
```

I've only tried this with a nested write that's two levels deep (e.g. query above). It might work as intended if only one level deep.

**Comment by Slmii** [comment]
I got the same issue, it only occurs with the following:

```prisma
generator client {
    provider        = "prisma-client-js"
    previewFeatures = ["uncheckedScalarInputs"]
}
```

```ts
const createActivities: Prisma.ActivityCreateInput[] | undefined = activities?.map(
	({ name, description, location, date, timezone, public: publicActivity, maxPeople, activityTypeId, transportationTypeId }) => ({
		name,
		description,
		location,
		date,
		timezone,
		public: publicActivity,
		maxPeople,
		trip: {},
		activityTypeId, // not working
		// activityType: { // working
		// 	connect: {
		// 		id: activityTypeId
		// 	}
		// }
	})
);

return prisma.trip.create({
	data: {
		...trip,
		userId,
		activities: {
			create: createActivities
		}
	}
});
```

using `"@prisma/cli": "^2.12.0",` & `"@prisma/client": "^2.12.0",`

**Reply by nikolasburk** [reply]
Hey @Slmii, thanks for adding this! Do you mind sharing your Prisma model definitions for `Activity` and `Trip` so I can try to reproduce this error?

**Reply by Slmii** [reply]
here it is @nikolasburk 

```prisma
model User {
    id        Int      @id @default(autoincrement())
    uuid      String   @default(uuid())
    email     String   @unique
    firstName String
    lastName  String
    password  String
    status    Boolean  @default(true)
    locked    Boolean  @default(false)
    role      Role     @default(USER)
    public    Boolean  @default(true)
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    trips      Trip[]
    activities UsersToActivities[]
}

model Trip {
    id            Int      @id @default(autoincrement())
    uuid          String   @default(uuid())
    userId        Int
    name          String
    description   String?
    public        Boolean
    dateFrom      DateTime
    dateTo        DateTime
    adults        Int
    children      Int?     @default(0)
    infants       Int?     @default(0)
    backgroundUrl String?
    createdAt     DateTime @default(now())
    updatedAt     DateTime @updatedAt

    user         User          @relation(fields: [userId], references: [id])
    activities   Activity[]

    @@unique([userId, id])
}

model Activity {
    id                   Int      @id @default(autoincrement())
    uuid                 String   @default(uuid())
    tripId               Int
    activityTypeId       Int
    name                 String
    description          String?
    location             String
    date                 DateTime
    timezone             String
    public               Boolean
    maxPeople            Int?     @default(0)
    createdAt            DateTime @default(now())
    updatedAt            DateTime @updatedAt

    trip               Trip                @relation(fields: [tripId], references: [id])
    activityType       ActivityType        @relation(fields: [activityTypeId], references: [id])
    users              UsersToActivities[]

    @@unique([tripId, id])
}

model ActivityType {
    id        Int      @id @default(autoincrement())
    uuid      String   @default(uuid())
    name      String
    type      String   @unique
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    activities Activity[]
}

model UsersToActivities {
    userId     Int
    activityId Int
    createdAt  DateTime @default(now())

    user     User     @relation(fields: [userId], references: [id])
    activity Activity @relation(fields: [activityId], references: [id])

    @@id([userId, activityId])
}
```

**Reply by nikolasburk** [reply]
@Slmii thanks a lot! To clarify: Are you using TypeScript and see a compiler error with this query or are you using plain JS and run into a runtime error?

**Reply by Slmii** [reply]
Yes, I am using TypeScript and I dont see any compile errors. Everything's fine. The error occurs when I run the mutation query with graphql. @nikolasburk 

Also, when I use `Prisma.ActivityCreateInput[]` it doesnt allow me to connect with an `id` , instead it expects an `activityType` with `connect`. Maybe I am using the wrong type here.

```js
 export type ActivityCreateInput = {
    ...
    activityType: ActivityTypeCreateOneWithoutActivitiesInput
  }

export type ActivityTypeCreateOneWithoutActivitiesInput = {
    create?: XOR<ActivityTypeCreateWithoutActivitiesInput, ActivityTypeUncheckedCreateWithoutActivitiesInput>
    connect?: ActivityTypeWhereUniqueInput
    connectOrCreate?: ActivityTypeCreateOrConnectWithoutactivitiesInput
  }
```
So I am trying now without typing `createActivities`, there error is still there tho

**Reply by nikolasburk** [reply]
Ah I see, thanks for clarifying!

> Maybe I am using the wrong type here.

I think this might actually be the issue. Can you try to use `Prisma.ActivityUncheckedCreateInput[]` instead of `Prisma.ActivityCreateInput[]` for `createActivities` instead and see if that solves the problem?

I'm a bit confused that you're seeing this error only at runtime though since these kinds of type errors _should_ be caught by the TS compiler 🤔

**Reply by Slmii** [reply]
so when using `ActivityUncheckedCreateInput` it doesnt give me a compile error anymore about using the `activityId` directly but the error still remains.

`const createActivities: Prisma.ActivityUncheckedCreateInput[] | undefined = activities?.map()`

could it be that its caused because `createActivities` can be undefined? The `activities` input data is not required.

The error btw:
![image](https://user-images.githubusercontent.com/9927849/100620386-aa0dd200-331e-11eb-8653-9688f3a8060b.png)

**Reply by nikolasburk** [reply]
Hmm yeah, I think that could be an issue... any way you can avoid this to be `undefined`?

**Reply by Slmii** [reply]
No, it's not a required relation (trip to many activities). It's an optional input data.

**Reply by Slmii** [reply]
Or did you mean for testing? Btw @nikolasburk when I was testing I always did with activities and not without, so it was never undefined when I tested.

**Reply by nikolasburk** [reply]
@Slmii Hmm no I think that's all right then! I actually raised this issue internally and it might be an issue on our end after all – would you mind reposting this as an actual GitHub [issue](https://github.com/prisma/prisma/issues/new) so that our engineers can look into it?

(Sorry for the inconvenience – unfortunately GitHub doesn't allow for transforming a discussion into an issue.)

**Reply by Slmii** [reply, marked as answer]
sorry for the late reply: https://github.com/prisma/prisma/issues/4442

**Comment by nikolasburk** [comment]
@pradyuman The `uncheckedScalarInputs` should work multiple levels deep, consider this working example:

```ts
const user = await prisma.user.create({
  data: {
    name: 'Alice',
    posts: {
      create: {
        title: 'Hello World',
        category: {
          create: {
            label: "Random",
            review: {
              create: {
                byUserId: 1
              }
            }
          }
        }
      },
    },
  },
})
```

Which is based on this sample schema:

```prisma
model User {
  id      Int      @id @default(autoincrement())
  name    String
  posts   Post[]
  Review  Review[]
}

model Post {
  id         Int      @id @default(autoincrement())
  title      String
  author     User     @relation(fields: [authorId], references: [id])
  authorId   Int // relation scalar field  (used in the `@relation` attribute above)
  categoryId Int
  category   Category @relation(fields: [categoryId], references: [id])
}

model Category {
  id     Int     @id @default(autoincrement())
  label  String
  posts  Post[]
  review Review?
}

model Review {
  id         Int      @id @default(autoincrement())
  categoryId Int
  category   Category @relation(fields: [categoryId], references: [id])
  byUserId   Int
  byUser     User     @relation(fields: [byUserId], references: [id])
}
```

It seems like there might be another issue with your query, since it didn't compile for me based on the Prisma schema you shared in the [other issue](https://github.com/prisma/prisma/discussions/4407).

![image](https://user-images.githubusercontent.com/4058327/100597266-85edc900-32fd-11eb-8137-79bce3d88c04.png)

I'm happy to dig further into this with you though! My general recommendation is to:

- make sure you have _all_ relation fields properly set in the Prisma schema (e.g. in the schema you shared in the issue before there were missing relation fields)
- make sure all required relations are specified when sending a `create` query with Prisma Client

**Reply by pradyuman** [reply]
I actually didn't cross reference this query with the one in the other post. This was just an example derived from the full schema I sent via DM for illustration purposes. Seems like @Slmii has a similar issue so I'm curious to know what the resolution is for his as well.

**Comment by louisabraham** [comment]
related to https://github.com/prisma/prisma/issues/4604

**Reply by pantharshit00** [reply]
I closed that one as a duplicate of https://github.com/prisma/prisma/issues/4442

This issue is confirmed and it will be addressed, hopefully in the next release.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
